### PR TITLE
Convert js squashed

### PR DESCRIPTION
**Pull request by Mersho**

### CODE REVIEW

File: bot/start.ts [hunk 1]
```diff
@@ -195,7 +195,7 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
         throw new Error(ctxUpdateAssertMsg);
       }
       const tgUser = ctx.update.message.from;
-      if (!tgUser.username) return await messages.nonHandleErrorMessage(ctx);
+      if (tgUser.username === undefined) return await messages.nonHandleErrorMessage(ctx);
```
Comment (knocte) [inline bot/start.ts]: @Mersho please don't change this line in the refactoring commit to eliminate ! usage when not boolean; I told you that this refactoring should ONLY affect lines that have been changed by the JS->TS conversion

File: bot/start.ts [hunk 2]
```diff
@@ -207,9 +207,9 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
   bot.command('maintenance', superAdminMiddleware, async (ctx: MainContext): Promise<void> => {
     try {
       const [val] = await validateParams(ctx, 2, '\\<_on/off_\\>');
-      if (!val) return;
+      if (val === null) return;
       let config = await Config.findOne();
-      if (!config) {
+      if (config === null) {
```
Comment (knocte) [inline bot/start.ts:213]: @Mersho are you absolutely sure that Config.findOne() can return null?

File: bot/start.ts [hunk 3]
```diff
@@ -327,6 +327,8 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
       order.canceled_by = ctx.admin._id;
       const buyer = await User.findOne({ _id: order.buyer_id });
       const seller = await User.findOne({ _id: order.seller_id });
+      if (buyer === null) return;
+      if (seller === null) return;
```
Comment (knocte) [inline bot/start.ts]: @Mersho why are you changing behaviour here?

File: bot/validations.ts [hunk 4]
```diff
@@ -1,18 +1,32 @@
...
-      : ctx.update.message.from;
+    let tgUser = null;
+    if (("callback_query" in ctx.update) && ctx.update.callback_query) {
+      tgUser = ctx.update.callback_query.from;
+    }
+    else if (("message" in ctx.update) && ctx.update.message) {
+      tgUser = ctx.update.message.from;
+    }
+    else {
+      throw new Error(ctxUpdateAssertMsg);
```
Comment (knocte) [inline bot/validations.ts:28]: @Mersho why are you changing behaviour here?

File: bot/validations.ts [hunk 5]
```diff
@@ -62,13 +79,16 @@ const validateSuperAdmin = async (ctx, id) => {
...
   try {
+    if (!('message' in ctx.update) || !('text' in ctx.update.message)) {
+      throw new Error(ctxUpdateAssertMsg);
+    }
     const tgUserId = id || ctx.update.message.from.id;
     const user = await User.findOne({ tg_id: tgUserId });
     // If the user never started the bot we can't send messages
     // to that user, so we do nothing
-    if (!user) return;
+    if (user == null) throw "uesr object is null";
```
Comment (knocte) [inline bot/validations.ts]: @Mersho typo here, but anyway, why are you changing behaviour here?

File: bot/validations.ts [hunk 6]
```diff
@@ -546,10 +566,13 @@ const validateSeller = async (ctx, user) => {
   }
 };
 
-const validateParams = async (ctx, paramNumber, errOutputString) => {
+const validateParams = async (ctx: MainContext, paramNumber: number, errOutputString: string): Promise<null | Array<string>> => {
   try {
+    if (!('message' in ctx.update) || !('text' in ctx.update.message)) {
```
Comment (knocte) [inline bot/validations.ts:573]: @Mersho why are you changing behaviour here?

File: bot/validations.ts [hunk 7]
```diff
@@ -562,11 +585,11 @@ const validateParams = async (ctx, paramNumber, errOutputString) => {
     return params.slice(1);
   } catch (error) {
     logger.error(error);
-    return false;
+    return null;
```
Comment (knocte) [inline bot/validations.ts:590]: @Mersho this is great; maybe this change (and the subsequente change over the callers of this func) should go in its own commit

File: db_connect.ts [hunk 8]
```diff
@@ -12,12 +12,13 @@ if (!MONGO_URI) {
   throw new Error('You must provide a MongoDB URI');
 }
 logger.info(`Connecting to: ${MONGO_URI}`);
+// TODO: Update mongoose to latest version
 const connect = () => {
   mongoose.connect(MONGO_URI, {
     useNewUrlParser: true,
     useUnifiedTopology: true,
-  });
+  } as any); // older version of mongoose that does not have the ConnectOptions type defined.
```
Comment (knocte) [inline db_connect.ts]: @Mersho let's update this in a separate commit, to avoid having to write this comment

File: bot/start.ts (hunk 2 above)
Comment (Mersho) [reply bot/start.ts:213]: yes

File: bot/start.ts (hunk 3 above)
Comment (Mersho) [reply bot/start.ts]: > @Mersho why are you changing behaviour here?

@knocte On lines 334 and 336, we received an error because buyer & seller might be null.

File: bot/start.ts (hunk 2 above)
Comment (knocte) [reply bot/start.ts:213]: @Mersho how about undefined?

File: bot/start.ts (hunk 3 above)
Comment (knocte) [reply bot/start.ts]: what error? you mean compiler error?

Comment (Mersho) [reply bot/start.ts]: I'm combining these two ifs

`if (buyer === null || seller === null) return;`

Comment (knocte) [reply bot/start.ts]: you didn't answer my last question

File: bot/validations.ts (hunk 6 above)
Comment (Mersho) [reply bot/validations.ts:573]: I saw Parham and Zahra doing similar things in their [previous PR](https://github.com/lnp2pBot/bot/commit/0da0893e327312efb6fb9fd80c0e75e3f2d9b3c3#diff-0b7b03c4bfeff599cc0bb0676e89f0a14d7ea0c01c0519d8c163acc0b20701dbR196).

We tried to add/change types and interfaces to the MainContext and it's not working.

File: bot/start.ts (hunk 2 above)
Comment (Mersho) [reply bot/start.ts:213]: no, the type of config is `IConfig | null` .

File: bot/start.ts (hunk 3 above)
Comment (Mersho) [reply bot/start.ts]: > what error? you mean compiler error?

yes, this is the error msg:

```
Argument of type '(UserDocument & { _id: ObjectId; }) | null' is not assignable to parameter of type 'UserDocument'.
  Type 'null' is not assignable to type 'UserDocument'.
```

File: bot/start.ts (hunk 2 above)
Comment (knocte) [reply bot/start.ts:213]: ok good

File: bot/validations.ts (hunk 4 above)
Comment (Mersho) [reply bot/validations.ts:28]: @knocte Here's the [same thing](https://github.com/nblockchain/bot/pull/1#discussion_r1324273457)

Comment (knocte) [reply bot/validations.ts:28]: same thing? your link points me to a comment which points me to Zara&Parham's commit, but in that commit there's no `throw` addition

Comment (knocte) [reply bot/validations.ts:28]: nevermind, the part that threw was in a file of a diff which was collapsed by default; however, in Zara&Parham's commit, the throw happens in an `if`, not in an `else`; let's not do it in an else, errors should be thrown ASAP, the higher in the function the better

Comment (Mersho) [reply bot/validations.ts:28]: > nevermind, the part that threw was in a file of a diff which was collapsed by default; however, in Zara&Parham's commit, the throw happens in an `if`, not in an `else`; let's not do it in an else, errors should be thrown ASAP, the higher in the function the better

If I remove the else statement, the function flow gets broken, so I don't know how to do that.

Comment (knocte) [reply bot/validations.ts:28]: I didn't ask you to remove the else statement, I asked you to put the throw in the `if` part instead of the `else` statement; to do that, you need to change the condition slightly

**Comment by knocte** [comment]
Please push 1by1

File: jobs/cancel_orders.ts [hunk 9]
```diff
@@ -45,22 +47,22 @@ const cancelOrders = async bot => {
     for (const order of activeOrders) {
       const buyerUser = await User.findOne({ _id: order.buyer_id });
       const sellerUser = await User.findOne({ _id: order.seller_id });
-      const i18nCtxBuyer = await getUserI18nContext(buyerUser);
-      const i18nCtxSeller = await getUserI18nContext(sellerUser);
+      const i18nCtxBuyer = await getUserI18nContext(buyerUser!);
+      const i18nCtxSeller = await getUserI18nContext(sellerUser!);
```
Comment (knocte) [inline jobs/cancel_orders.ts]: @Mersho can you explain this please? /cc @aarani

File: util/index.ts [hunk 10]
```diff
@@ -6,29 +14,29 @@ const { Order, Community } = require('../models');
...
+  return code.split('').every(letter => {
     if (alphabet.indexOf(letter) == -1) {
       return false;
     }
     return true;
   });
 };
 
-const getCurrency = code => {
+const getCurrency = (code: string): (IFiat | boolean) => {
```
Comment (knocte) [inline util/index.ts]: @Mersho this looks like same bad practice as with validateParams? if yes, let's fix it in the same commit where we fix that

File: util/index.ts [hunk 11]
```diff
@@ -184,27 +195,27 @@ const decimalRound = (value, exp) => {
...
+  value = Math.round(+(valueArr[0] + 'e' + (valueArr[1] ? +valueArr[1] - exp : -exp)));
   // Shift back
-  value = value.toString().split('e');
-  return +(value[0] + 'e' + (value[1] ? +value[1] + exp : exp));
+  valueArr = value.toString().split('e');
+  return +(valueArr[0] + 'e' + (valueArr[1] ? +valueArr[1] + exp : exp));
 };
 
-const extractId = text => {
+const extractId = (text: string): string | undefined => {
```
Comment (knocte) [inline util/index.ts]: @Mersho please don't violate our WorkflowGuidelines

File: jobs/cancel_orders.ts (hunk 9 above)
Comment (Mersho) [reply jobs/cancel_orders.ts]: > @Mersho can you explain this please? /cc @aarani

I'll fix it. The null-check I did previously with this operator was completely incorrect.

Comment (Mersho) [reply jobs/cancel_orders.ts]: I should throw, but I am changing this like the original repo

`if (buyerUser === null || sellerUser === null) return;`

File: util/index.ts (hunk 11 above)
Comment (Mersho) [reply util/index.ts]: @knocte Done. If there's anything wrong, let me know.

File: util/index.ts (hunk 10 above)
Comment (Mersho) [reply util/index.ts]: @knocte Done. If there's anything wrong, let me know.

Comment (Mersho) [reply util/index.ts]: oh wait wait

Comment (Mersho) [reply util/index.ts]: It's done.

File: jobs/cancel_orders.ts (hunk 9 above)
Comment (knocte) [reply jobs/cancel_orders.ts]: wouldnt that change behaviour? /cc @aarani

**Comment by knocte** [comment]
Even though we've had good progress here, I still see many issues. And it feels that the issues are new (they were not there before); so it seems that @Mersho, you're creating new issues when addressing the issues I have pointed out before.

To avoid this happening again, we're going to do the following:
1. I'm going to point out now the new issues I see, but please don't work on them yet.
2. Then, tomorrow, we do pair-programming on them (in jitsi). I'll be watching you address each issue so that you don't create new issues when addressing them.
3. After we finish our session, we can close this PR and open a new one in the upstream repo.

**Comment by knocte** [comment]
Issues in commit messages:

> We need to enable esModuleInterop for SocksProxyAgent
> because this solves error TS1192.

Please add a link to that TS1192 error.


> The older version of mongoose does not have
> the ConnectOptions type defined. make sure
> to update mongoose to latest version later.

Two issues above:
* In the English language, the next word after a dot has to start with uppercase.
* After fixing the above, you seem to be telling an order to some one "please make sure to update mongoose to latest version later". An order in a commit message? Orders should go in github issues or TODO/FIXME comments.
* Even if you move that "order" to a TODO comment, I told you in a previous review that I want this issue to be solved in a commit. I want a new commit that upgrades mongoose to the latest version.


> Also search/replace from require to import &
> Replacing "module.exports" with "export".

The commit message body is composed of paragraphs. Each paragraph ends with a dot, and each paragraph starts with an uppercase letter. However, in the above message, I seem to see that a second paragraph starts with an uppercase word ("Replacing"), however, I don't see a dot before this word; so it is very confusing.

File: app.ts [hunk 12]
```diff
@@ -30,12 +31,13 @@ const { delay } = require('./util');
           telegram: {
             agent,
           },
-        };
+        } as any;
```
Comment (knocte) [inline app.ts]: @Mersho I told you before that I want this issue fixed in this PR

File: bot/start.ts [hunk 13]
```diff
@@ -256,11 +256,11 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
         throw new Error(ctxUpdateAssertMsg);
       }
       const params = ctx.update.message.text.split(' ');
-      const [command, orderId] = params.filter((el) => el);
+      const [command, orderId] = params.filter((el: string) => el);
 
       if (!orderId) {
         const orders = await askForConfirmation(ctx.user, command);
-        if (!orders.length) return await ctx.reply(`${command} <order Id>`);
+        if (!orders?.length) return await ctx.reply(`${command} <order Id>`);
```
Comment (knocte) [inline bot/start.ts]: @Mersho I have the feeling that we can improve the above ^

File: bot/start.ts [hunk 14]
```diff
@@ -327,6 +327,7 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
       order.canceled_by = ctx.admin._id;
       const buyer = await User.findOne({ _id: order.buyer_id });
       const seller = await User.findOne({ _id: order.seller_id });
+      if (buyer === null || seller === null) return;
```
Comment (knocte) [inline bot/start.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: bot/start.ts [hunk 15]
```diff
@@ -347,11 +348,11 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
         throw new Error(ctxUpdateAssertMsg);
       }
       const params = ctx.update.message.text.split(' ');
-      const [command, orderId] = params.filter((el) => el);
+      const [command, orderId] = params.filter((el: string) => el);
 
       if (!orderId) {
         const orders = await askForConfirmation(ctx.user, command);
-        if (!orders.length) return await ctx.reply(`${command}  <order Id>`);
+        if (!orders?.length) return await ctx.reply(`${command}  <order Id>`);
```
Comment (knocte) [inline bot/start.ts]: @Mersho same here

File: bot/start.ts [hunk 16]
```diff
@@ -426,6 +427,8 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
       order.status = 'COMPLETED_BY_ADMIN';
       const buyer = await User.findOne({ _id: order.buyer_id });
       const seller = await User.findOne({ _id: order.seller_id });
+      if (buyer === null) return;
+      if (seller === null) return;
```
Comment (knocte) [inline bot/start.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: bot/start.ts [hunk 17]
```diff
@@ -450,10 +453,12 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
       if (!(await validateObjectId(ctx, orderId))) return;
       const order = await Order.findOne({ _id: orderId });
 
-      if (!order) return;
+      if (order === null) return;
 
       const buyer = await User.findOne({ _id: order.buyer_id });
       const seller = await User.findOne({ _id: order.seller_id });
+      if (buyer === null) return;
+      if (seller === null) return;
```
Comment (knocte) [inline bot/start.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: bot/start.ts [hunk 18]
```diff
@@ -523,11 +528,11 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
         throw new Error(ctxUpdateAssertMsg);
       }
       const params = ctx.update.message.text.split(' ');
-      const [command, orderId] = params.filter((el) => el);
+      const [command, orderId] = params.filter((el: string) => el);
 
       if (!orderId) {
         const orders = await askForConfirmation(ctx.user, command);
-        if (!orders.length) return await ctx.reply(`${command} <order Id>`);
+        if (!orders?.length) return await ctx.reply(`${command} <order Id>`);
```
Comment (knocte) [inline bot/start.ts]: @Mersho same here

File: bot/start.ts [hunk 19]
```diff
@@ -565,11 +570,11 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
           const community = await Community.findById(
             ctx.admin.default_community_id
           );
-          community.banned_users.push({
+          community?.banned_users.push({
```
Comment (knocte) [inline bot/start.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: bot/start.ts [hunk 20]
```diff
@@ -565,11 +570,11 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
...
             id: user._id,
             username: user.username,
           });
-          await community.save();
+          await community?.save();
```
Comment (knocte) [inline bot/start.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: bot/start.ts [hunk 21]
```diff
@@ -608,8 +613,9 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
           const community = await Community.findById(
             ctx.admin.default_community_id
           );
-          community.banned_users = community.banned_users.filter(
-            (el: any) => el.id !== user.id
+          if (community === null) return;
```
Comment (knocte) [inline bot/start.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: bot/validations.ts [hunk 22]
```diff
@@ -306,20 +326,20 @@ const validateInvoice = async (ctx, lnInvoice) => {
...
-const isValidInvoice = async (ctx, lnInvoice) => {
+const isValidInvoice = async (ctx: MainContext, lnInvoice: string) => {
   try {
     const invoice = parsePaymentRequest({ request: lnInvoice });
     const latestDate = new Date(
-      Date.now() + parseInt(process.env.INVOICE_EXPIRATION_WINDOW)
+      Date.now() + Number(process.env.INVOICE_EXPIRATION_WINDOW)
     ).toISOString();
-    if (!!invoice.tokens && invoice.tokens < process.env.MIN_PAYMENT_AMT) {
+    if (!!invoice.tokens && invoice.tokens < Number(process.env.MIN_PAYMENT_AMT)) {
```
Comment (knocte) [inline bot/validations.ts:337]: @Mersho this looks like a bug that is existing upstream? we should call parseFloat() to fix the bug first AFAIU

File: db_connect.ts [hunk 23]
```diff
@@ -16,8 +16,8 @@ const connect = () => {
   mongoose.connect(MONGO_URI, {
     useNewUrlParser: true,
     useUnifiedTopology: true,
-  });
+  } as any);
```
Comment (knocte) [inline db_connect.ts]: @Mersho same here

File: jobs/node_info.ts [hunk 24]
```diff
@@ -1,20 +1,24 @@
...
+import { Config } from '../models';
 const { getInfo } = require('../ln');
-const logger = require('../logger');
+import logger from "../logger";
 
-const info = async bot => {
+const info = async (bot: Telegraf<MainContext>) => {
   try {
     const config = await Config.findOne({});
+    if (config === null) return;
```
Comment (knocte) [inline jobs/node_info.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: jobs/pending_payments.ts [hunk 25]
```diff
@@ -16,6 +16,7 @@ exports.attemptPendingPayments = async (bot: Telegraf<MainContext>): Promise<voi
   });
   for (const pending of pendingPayments) {
     const order = await Order.findOne({ _id: pending.order_id });
+    if (order == null) throw "order object is null";
```
Comment (knocte) [inline jobs/pending_payments.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: jobs/pending_payments.ts [hunk 26]
```diff
@@ -38,6 +39,7 @@ exports.attemptPendingPayments = async (bot: Telegraf<MainContext>): Promise<voi
         request: pending.payment_request,
       });
       const buyerUser = await User.findOne({ _id: order.buyer_id });
+      if (buyerUser === null) return;
```
Comment (knocte) [inline jobs/pending_payments.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: jobs/pending_payments.ts [hunk 27]
```diff
@@ -55,12 +57,13 @@ exports.attemptPendingPayments = async (bot: Telegraf<MainContext>): Promise<voi
...
         order.routing_fee = payment.fee;
         pending.paid = true;
-        pending.paid_at = new Date().toISOString();
+        pending.paid_at = new Date();
         // We add a new completed trade for the buyer
         buyerUser.trades_completed++;
         await buyerUser.save();
         // We add a new completed trade for the seller
         const sellerUser = await User.findOne({ _id: order.seller_id });
+        if (sellerUser === null) return;
```
Comment (knocte) [inline jobs/pending_payments.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: jobs/pending_payments.ts [hunk 28]
```diff
@@ -134,6 +137,7 @@ exports.attemptCommunitiesPendingPayments = async (bot: Telegraf<MainContext>):
         request: pending.payment_request,
       });
       const user = await User.findById(pending.user_id);
+      if (user === null) return;
```
Comment (knocte) [inline jobs/pending_payments.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: jobs/pending_payments.ts [hunk 29]
```diff
@@ -145,9 +149,10 @@ exports.attemptCommunitiesPendingPayments = async (bot: Telegraf<MainContext>):
       }
 
       const community = await Community.findById(pending.community_id);
+      if (community === null) return;
```
Comment (knocte) [inline jobs/pending_payments.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: ln/subscribe_invoice.ts [hunk 30]
```diff
@@ -1,23 +1,27 @@
...
+import logger from "../logger";
 
-const subscribeInvoice = async (bot, id, resub) => {
+const subscribeInvoice = async (bot: Telegraf<MainContext>, id: string, resub: boolean) => {
   try {
     const sub = subscribeToInvoice({ id, lnd });
     sub.on('invoice_updated', async invoice => {
       if (invoice.is_held && !resub) {
         const order = await Order.findOne({ hash: invoice.id });
+        if (order === null) return;
```
Comment (knocte) [inline ln/subscribe_invoice.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: ln/subscribe_invoice.ts [hunk 31]
```diff
@@ -1,23 +1,27 @@
...
         logger.info(
           `Order ${order._id} Invoice with hash: ${id} is being held!`
         );
         const buyerUser = await User.findOne({ _id: order.buyer_id });
         const sellerUser = await User.findOne({ _id: order.seller_id });
+        if (buyerUser === null || sellerUser === null) return;
```
Comment (knocte) [inline ln/subscribe_invoice.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: ln/subscribe_invoice.ts [hunk 32]
```diff
@@ -47,18 +51,20 @@ const subscribeInvoice = async (bot, id, resub) => {
             rate
           );
         }
-        order.invoice_held_at = Date.now();
+        order.invoice_held_at = new Date();
         order.save();
       }
       if (invoice.is_confirmed) {
         const order = await Order.findOne({ hash: id });
+        if (order === null) return;
```
Comment (knocte) [inline ln/subscribe_invoice.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: ln/subscribe_invoice.ts [hunk 33]
```diff
@@ -47,18 +51,20 @@ const subscribeInvoice = async (bot, id, resub) => {
...
         logger.info(
           `Order ${order._id} - Invoice with hash: ${id} was settled!`
         );
         order.status = 'PAID_HOLD_INVOICE';
         await order.save();
         const buyerUser = await User.findOne({ _id: order.buyer_id });
         const sellerUser = await User.findOne({ _id: order.seller_id });
+        if (buyerUser === null || sellerUser === null) return;
```
Comment (knocte) [inline ln/subscribe_invoice.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: lnurl/lnurl-pay.ts [hunk 34]
```diff
@@ -17,7 +17,7 @@ const resolvLightningAddress = async (address, amountMsat) => {
   }
 
   if (
-    (lnAddressRes.minSendable > amountMsat) |
+    (lnAddressRes.minSendable > amountMsat) ||
```
Comment (knocte) [inline lnurl/lnurl-pay.ts:20]: @Mersho in JavaScript a single `|` char means OR?

File: util/index.ts [hunk 35]
```diff
@@ -6,29 +14,29 @@ const { Order, Community } = require('../models');
...
   });
 };
 
-const getCurrency = code => {
-  if (!isIso4217(code)) return false;
+const getCurrency = (code: string): (IFiat | null) => {
+  if (!isIso4217(code)) return null;
   const currency = currencies[code];
-  if (!currency) return false;
+  if (currency === undefined) return null;
```
Comment (knocte) [inline util/index.ts]: @Mersho please do not violate our workflow guidelines

File: util/index.ts [hunk 36]
```diff
@@ -118,9 +126,10 @@ const handleReputationItems = async (buyer, seller, amount) => {
   }
 };
 
-const getBtcFiatPrice = async (fiatCode, fiatAmount) => {
+const getBtcFiatPrice = async (fiatCode: string, fiatAmount: number) => {
   try {
     const currency = getCurrency(fiatCode);
+    if (!(currency instanceof Object)) return;
```
Comment (knocte) [inline util/index.ts]: @Mersho why are you changing behaviour here? converting from JS to TS should not change the behaviour of the program

File: util/index.ts [hunk 37]
```diff
@@ -147,22 +156,23 @@ const getBtcExchangePrice = (fiatAmount, satsAmount) => {
...
-  const array = [];
+const objectToArray = (object: any): any[] => {
+  const array: any[] = [];
 
   for (const i in object) array.push(object[i]);
 
   return array;
 };
 
+
```
Comment (knocte) [inline util/index.ts]: @Mersho let's get rid of unnecessary whitespace changes

File: util/index.ts [hunk 38]
```diff
@@ -184,27 +194,29 @@ const decimalRound = (value, exp) => {
...
+  valueArr = value.toString().split('e');
+  return +(valueArr[0] + 'e' + (valueArr[1] ? +valueArr[1] + exp : exp));
 };
 
-const extractId = text => {
+const extractId = (text: string): (string | null) => {
   const matches = text.match(/:([a-f0-9]{24}):$/);
-
-  return matches[1];
+  if (matches){
```
Comment (knocte) [inline util/index.ts]: @Mersho is matches a boolean? if not, then don't violate our workflow guidelines

File: util/index.ts [hunk 39]
```diff
@@ -307,8 +319,10 @@ const getDisputeChannel = async order => {
  * @param {*} user
  * @returns i18n context
  */
-const getUserI18nContext = async user => {
-  const language = user.language || 'en';
+const getUserI18nContext = async (user: UserDocument) => {
+  const language = user.lang || 'en';  // UserDocument doesn't have language attr
+  // I think language always be 'en'
+  // BTW I fixed this!
```
Comment (knocte) [inline util/index.ts]: @Mersho what is going on here? this comments is completely confusing for future developers when they work on this file

File: util/index.ts [hunk 40]
```diff
@@ -307,8 +319,10 @@ const getDisputeChannel = async order => {
  * @param {*} user
  * @returns i18n context
  */
-const getUserI18nContext = async user => {
-  const language = user.language || 'en';
+const getUserI18nContext = async (user: UserDocument) => {
+  const language = user.lang || 'en';  // UserDocument doesn't have language attr
```
Comment (knocte) [inline util/index.ts]: @Mersho please don't violate our workflow guidelines

File: util/index.ts [hunk 41]
```diff
@@ -371,55 +385,55 @@ const isDisputeSolver = (community, user) => {
...
+const itemsFromMessage = (str: string) => {
   return str
     .split(' ')
     .map(e => e.trim())
     .filter(e => !!e);
 };
 
 // Check if a number is int
-const isInt = n => parseInt(n) === n;
+const isInt = (n: number) => Number(n) === n;
```
Comment (knocte) [inline util/index.ts]: @Mersho this looks like a function that won't be needed when all code is TypeScript

File: util/index.ts [hunk 42]
```diff
@@ -434,13 +448,13 @@ const getUserAge = user => {
...
-const getTimeToExpirationOrder = (order, i18n) => {
+const getTimeToExpirationOrder = (order: IOrder, i18n: I18nContext) => {
   const initialDateObj = new Date(order.created_at);
-  const timeToExpire = parseInt(process.env.ORDER_PUBLISHED_EXPIRATION_WINDOW);
+  const timeToExpire = Number(process.env.ORDER_PUBLISHED_EXPIRATION_WINDOW);
   initialDateObj.setSeconds(initialDateObj.getSeconds() + timeToExpire);
 
   const currentDateObj = new Date();
-  const timeDifferenceMs = initialDateObj - currentDateObj;
+  const timeDifferenceMs = initialDateObj.getTime() - currentDateObj.getTime();
```
Comment (knocte) [inline util/index.ts]: @Mersho are you sure you're not introducing a bug here?

File: util/index.ts (hunk 37 above)
Comment (Mersho) [reply util/index.ts]: Done.

File: bot/validations.ts [hunk 43]
```diff
@@ -259,26 +281,27 @@ const validateBuyOrder = async ctx => {
...
-const validateInvoice = async (ctx, lnInvoice) => {
+const validateInvoice = async (ctx: MainContext, lnInvoice: string) => {
   try {
     const invoice = parsePaymentRequest({ request: lnInvoice });
     const latestDate = new Date(
-      Date.now() + parseInt(process.env.INVOICE_EXPIRATION_WINDOW)
+      Date.now() + Number(process.env.INVOICE_EXPIRATION_WINDOW)
     ).toISOString();
-    if (!!invoice.tokens && invoice.tokens < process.env.MIN_PAYMENT_AMT) {
+    if (process.env.MIN_PAYMENT_AMT === undefined) throw Error("MIN_PAYMENT_AMT is undefined, please check .env file");
```
Comment (knocte) [inline bot/validations.ts]: @Mersho to avoid using the word `undefined` here (to respect our workflow guidelines), I guess you can do `if ("MAIN_PAYMENT_AMT" in process.env)`?

File: util/fiatModel.ts [hunk 44]
```diff
@@ -10,3 +10,7 @@ export interface IFiat {
     price?: boolean;
     locale?: string;
 }
+
+export interface IFiats {
```
Comment (knocte) [inline util/fiatModel.ts]: @Mersho better name: IFiatCurrencies

Comment (Mersho) [reply util/fiatModel.ts]: Done

File: bot/validations.ts (hunk 43 above)
Comment (Mersho) [reply bot/validations.ts]: Done

**Comment by knocte** [comment]
Last but not least, this commit title:

```
style: fix/active eslint on typescript
```

fix/active? I think there's a typo there, otherwise can't parse.

**Comment by Mersho** [comment]
> Last but not least, this commit title:
> 
> ```
> style: fix/active eslint on typescript
> ```
> 
> fix/active? I think there's a typo there, otherwise can't parse.

Yes, you're right, I changed the commit message.

**Comment by knocte** [comment]
>  I changed the commit message.

I see same issue. Active is not a verb, it's an adjective.

**Comment by Mersho** [comment]
> > I changed the commit message.
> 
> I see same issue. Active is not a verb, it's an adjective.

Commit msg edited again. let me know if there's any thing wrong.

**Comment by knocte** [comment]
I guess you were looking for the verb **activate**. But enable sounds even better.
